### PR TITLE
Loans: multi cashflows fix external loan

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

External pricing loan with correct cashflow computation.

For computing interest of:
- internal loans: we use `debt - principal`
- external loans: we use `debt - notional`

From the cashflow perspective, we can emulate the `debt` at maturity given the `interest` and `principal` parameters given to the `expected_cashflows()`. Nevertheless, in the case of external loans, the principal used for computing the debt is different from the value substracted to the debt to obtain the interest. We need to pass a new parameter to `expected_cashflows()`. Called `principal_base`, which is the unaltered principal. It maps to `principal` for internal pricing and `notional` for external pricing.

### CODE REVIEW

File: pallets/loans/src/types/cashflow.rs [hunk 1]
```diff
@@ -173,11 +171,7 @@ impl RepaymentSchedule {
 			.map(|(_, weight)| weight)
 			.try_fold(0, |a, b| a.ensure_add(*b))?;
 
-		let lifetime = maturity.ensure_sub(origination_date)?.ensure_into()?;
-		let interest_rate_per_lifetime = ensure_pow(interest_rate.per_sec()?, lifetime)?;
-		let interest_at_maturity = interest_rate_per_lifetime
-			.ensure_mul_int(principal)?
-			.ensure_sub(principal)?;
+		let interest_at_maturity = current_debt.ensure_sub(principal)?;
```
Comment (lemunozm) [inline pallets/loans/src/types/cashflow.rs:174]: Basically, the principal is different depending on the kind of pricing. The current debt is what previously it was:

```rust
let lifetime = maturity.ensure_sub(origination_date)?.ensure_into()?;
let interest_rate_per_lifetime = ensure_pow(interest_rate.per_sec()?, lifetime)?;
let interest_at_maturity = interest_rate_per_lifetime
	.ensure_mul_int(principal)?
```

We can do this simplification now due it is computed as paid at maturity

File: pallets/loans/src/entities/loans.rs [hunk 2]
```diff
@@ -243,26 +243,21 @@ impl<T: Config> ActiveLoan<T> {
...
-				.ensure_sub(self.total_repaid.principal)?),
-			ActivePricing::External(inner) => {
-				let maturity = self.maturity_date();
-				inner.outstanding_principal(pool_id, maturity)
-			}
-		}
+	pub fn principal(&self) -> Result<T::Balance, DispatchError> {
+		Ok(self
+			.total_borrowed
+			.ensure_sub(self.total_repaid.principal)?)
```
Comment (mustermeiszer) [inline pallets/loans/src/entities/loans.rs]: Why the change? Does this mean, that the gains we are making through price increases are repayed as unscheduled?

Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs]: It seems like you have reviewed something outdated

Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs]: But I get your point, it should be just the borrowed amount, not taking into account the repaid one, right?

Comment (mustermeiszer) [reply pallets/loans/src/entities/loans.rs]: Now its compeltly gone, right? But looks good now

Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs]: It's still there, I need to remove the `.ensure_sub(self.total_repaid.principal)?` part, to not reduce the cashflow expected amount with the paid amount

**Comment by lemunozm** [comment]
I'll merge anyway both PRs into the same one due you already have reviewed this, and fixing this in the main one.
